provider/traceutils.go

```diff
@@ -0,0 +1,258 @@
+// Copyright 2016-2023, Pulumi Corporation.  All rights reserved.
+
+package gcp
+
+import (
+	"bufio"
+	//"bytes"
+	"encoding/csv"
+	"fmt"
+	"io"
+	"os"
+	"path"
+	"strings"
+	"time"
+
+	"sourcegraph.com/sourcegraph/appdash"
+)
+
+// tools for working with trace spans
+// TODO(pulumi/home#3168): The functions here should be generalized to live in pulumi/pulumi-trace-tool
+
+const span_id_key = "Span.ID"
+const span_start_key = "Span.Start"
+const span_end_key = "Span.End"
+const span_filename_key = "filename"
+
+// Convenience type for attaching helpers to string map that represents a span
+type Span map[string]string
+
+func (s Span) getId() (*appdash.SpanID, error) {
+	spanIdStr, ok := s[span_id_key]
+	if !ok {
+		return nil, fmt.Errorf("'%s' Not found", span_id_key)
+	}
+
+	spanId, err := appdash.ParseSpanID(spanIdStr)
+	if err != nil {
+		return nil, err
+	}
+	return spanId, nil
+}
+
+func (s Span) Start() (time.Time, error) {
+	spanStart, err := parseTime(s[span_start_key])
+	if err != nil {
+		err = fmt.Errorf("Failed to parse %s time: %w", span_start_key, err)
+		return time.Time{}, err
+	}
+	return spanStart, nil
+}
+
+func (s Span) End() (time.Time, error) {
+	spanEnd, err := parseTime(s[span_end_key])
+	if err != nil {
+		err = fmt.Errorf("Failed to parse %s time: %w", span_end_key, err)
+		return time.Time{}, err
+	}
+	return spanEnd, nil
+}
+
+func (s Span) Duration() (time.Duration, error) {
+	spanStart, err := s.Start()
+	if err != nil {
+		return 0, err
+	}
+	spanEnd, err := s.End()
+	if err != nil {
+		return 0, err
+	}
+	dur := spanEnd.Sub(spanStart)
+	return dur, nil
+}
+
+func parseTime(str string) (time.Time, error) {
+	return time.Parse(time.RFC3339, str)
+}
+
+// Interface to generalize iterating over all spans from various trace formats
+type SpanIter interface {
+	WalkSpans(onSpan func(x Span) error) error
+}
+
+
+// Convenience type for iterating over multiple SpanIters in sequence
+type MultiSpanIter struct {
+	Iters []SpanIter
+}
+
+var _ = SpanIter(MultiSpanIter{})
+
+func (i MultiSpanIter) WalkSpans(onSpan func(s Span) error) error {
+	for _, iter := range i.Iters {
+		err := iter.WalkSpans(onSpan)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+// Iterates over spans stored in CSV format
+type CsvFileSpanIter struct {
+	File string
+}
+
+var _ = SpanIter(CsvFileSpanIter{})
```

Comment:
Does this ensure the struct conforms to the interface?

Reply:
It looks like that's the intent. The form I use these days is `var _ SpanIter = CsvFileSpanIter{}` is probably equivalent.

Reply:
Yeah this trick threw me for a loop the first time I saw it too. Why have an explicit keyword for asserting compliance with an interface when you can just abuse casting apparently?